### PR TITLE
RateMSEDistortionLoss layer

### LINKED ISSUE #133: Rate distortion loss
## Enhancement

A rate distortion loss 

## Motivation

Rate distortion losses are used by a variety of models

### PR DESCRIPTION
**Pull request by 0x00b1**
Closes #133.

## Changes

- [x] `RateMSEDistortionLoss` layer

## Testing

Unit test

### CODE REVIEW

**Comment by 0x00b1** [comment]
@mmuckley Nothing too fancy.

File: neuralcompression/layers/_rate_distortion_loss.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+import torch
+from torch import Tensor
+from torch.nn import Module, MSELoss
+
+
+class _ForwardReturnType(NamedTuple):
+    bpp: float
+    mse: float
+    rate_distortion: float
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_distortion_loss.py:17]: Could we make this public? And maybe change the name to `RDLoss` or something.

File: neuralcompression/layers/_rate_distortion_loss.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class RateDistortionLoss(Module):
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_distortion_loss.py:20]: There are many definitions of distortion, but this function only covers MSE. So maybe we should change the name to `RateMSELoss`.

File: neuralcompression/layers/_rate_distortion_loss.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    """Rate-distortion loss.
+
+    The rate-distortion loss is the minimum transmission bit-rate for a
+    required quality. It can be obtained without consideration of a specific
+    coding method.
+
+    Args:
+        _lambda: `_lambda = 1` is the solution where the `(rate, distortion)`
+            pair minimizes `rate + distortion`. Increasing `_lambda` will
+            penalize the distortion term so more bits are spent.
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_distortion_loss.py:30]: This doesn't need to be private does it?

File: neuralcompression/layers/_rate_distortion_loss.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        self._lambda = _lambda
+
+        self._mse = MSELoss()
+
+    def forward(
+        self,
+        x_hat: Tensor,
+        scores: List[Tensor],
+        x: Tensor,
+    ) -> _ForwardReturnType:
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_distortion_loss.py:45]: Need a docstring describing what `scores` is.

File: neuralcompression/layers/_rate_distortion_loss.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+            pixels = -math.log(2) * (n * h * w)
+
+            bpps += [float(bits / pixels)]
+
+        bpp = sum(bpps)
+
+        mse = self._mse(x_hat, x)
+
+        rate_distortion = self._lambda * 255 ** 2 * mse + bpp
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_distortion_loss.py:61]: 255 seems a reasonable scaling factor but may not be universal. Could this be a parameter?

**Review by mmuckley — CHANGES_REQUESTED** [review]
Requesting a few changes, almost all related to function names to correctly indicate the scope to the user.

File: neuralcompression/layers/_rate_distortion_loss.py (hunk 2 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_distortion_loss.py]: Great idea. Changed.

File: neuralcompression/layers/_rate_distortion_loss.py (hunk 1 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_distortion_loss.py]: Yep. I refactored this a bit to make it more accessible.

File: neuralcompression/layers/_rate_distortion_loss.py (hunk 5 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_distortion_loss.py]: Yep! Updated.

File: neuralcompression/layers/_rate_distortion_loss.py (hunk 3 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_distortion_loss.py]: `lambda` is reserved. I went ahead and renamed `lambda` to `trade_off`.

File: neuralcompression/layers/_rate_distortion_loss.py (hunk 4 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_distortion_loss.py]: Updated. I also renamed `scores` to the clearer `bits`.

**Comment by 0x00b1** [comment]
Amazing review @mmuckley. You helped make this much nicer.

File: neuralcompression/layers/_rate_mse_distortion_loss.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class RateDistortionLoss(NamedTuple):
+    distortion: float
+    rate: float
+    rate_distortion: float
+
+
+class RateMSEDistortionLoss(Module):
+    """Rate-distortion loss.
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_mse_distortion_loss.py:21]: Missing MSE.

File: neuralcompression/layers/_rate_mse_distortion_loss.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    ``x``, distortion is expressed as the mean squared error (MSE) between the
+    original, ``x``, and the target, ``x_hat``.
+
+    Args:
+        trade_off: rate-distortion trade-off. `trade_off = 1` is the solution
+            where the `(rate, distortion)` pair minimizes `rate + distortion`.
+            Increasing `trade_off` will penalize the distortion term so more
+            bits are spent.
+        maximum: dynamic range of the input (i.e. the difference between the
+            maximum the and minimum permitted values).
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_mse_distortion_loss.py:35]: Great name.

File: neuralcompression/layers/_rate_mse_distortion_loss.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    def forward(
+        self,
+        x_hat: Tensor,
+        bits: List[Tensor],
+        x: Tensor,
+    ) -> RateDistortionLoss:
+        """
+        Args:
+            x_hat: encoder output.
+            bits: predicted bits.
```
Comment (mmuckley) [inline neuralcompression/layers/_rate_mse_distortion_loss.py:56]: Maybe a little bit more for what these are, relating to probabilities. Also you make a `log(2)` conversion below so are they actually bits?

**Review by mmuckley — APPROVED** [review]
LGTM. Nice update!

File: neuralcompression/layers/_rate_mse_distortion_loss.py (hunk 7 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_mse_distortion_loss.py:35]: Thanks @mmuckley :)

File: neuralcompression/layers/_rate_mse_distortion_loss.py (hunk 8 above)
Comment (0x00b1) [reply neuralcompression/layers/_rate_mse_distortion_loss.py]: Good point. Addressed.
